tf_agents/agents/dqn/examples/v1/train_eval_rnd.py

```diff
@@ -0,0 +1,306 @@
+      rb_checkpointer.initialize_or_restore(sess)
+      sess.run(iterator.initializer)
+      common.initialize_uninitialized_variables(sess)
+
+      sess.run(init_agent_op)
+      sess.run(train_summary_writer.init())
+      sess.run(eval_summary_writer.init())
+      sess.run(initial_collect_op)
+
+      global_step_val = sess.run(global_step)
+      metric_utils.compute_summaries(
+          eval_metrics,
+          eval_py_env,
+          eval_py_policy,
+          num_episodes=num_eval_episodes,
+          global_step=global_step_val,
+          callback=eval_metrics_callback,
+          log=True,
+      )
+
+      collect_call = sess.make_callable(collect_op)
+      global_step_call = sess.make_callable(global_step)
+      train_step_call = sess.make_callable([train_op, summary_ops])
+
+      timed_at_step = global_step_call()
+      collect_time = 0
+      train_time = 0
+      steps_per_second_ph = tf.compat.v1.placeholder(
+          tf.float32, shape=(), name='steps_per_sec_ph')
+      steps_per_second_summary = tf.compat.v2.summary.scalar(
+          name='global_steps_per_sec', data=steps_per_second_ph,
+          step=global_step)
+
+      for _ in range(num_iterations):
+        # Train/collect/eval.
+        start_time = time.time()
+        collect_call()
+        collect_time += time.time() - start_time
+        start_time = time.time()
+        for _ in range(train_steps_per_iteration):
+          loss_info_value, _ = train_step_call()
+        train_time += time.time() - start_time
+
+        global_step_val = global_step_call()
+
+        if global_step_val % log_interval == 0:
+          logging.info('step = %d, loss = %f', global_step_val,
+                       loss_info_value.loss)
+          steps_per_sec = (
+              (global_step_val - timed_at_step) / (collect_time + train_time))
+          sess.run(
+              steps_per_second_summary,
+              feed_dict={steps_per_second_ph: steps_per_sec})
+          logging.info('%.3f steps/sec', steps_per_sec)
+          logging.info('%s', 'collect_time = {}, train_time = {}'.format(
+              collect_time, train_time))
+          timed_at_step = global_step_val
+          collect_time = 0
+          train_time = 0
+
+        if global_step_val % train_checkpoint_interval == 0:
+          train_checkpointer.save(global_step=global_step_val)
+
+        if global_step_val % policy_checkpoint_interval == 0:
+          policy_checkpointer.save(global_step=global_step_val)
+
+        if global_step_val % rb_checkpoint_interval == 0:
+          rb_checkpointer.save(global_step=global_step_val)
+
+        if global_step_val % eval_interval == 0:
+          metric_utils.compute_summaries(
+              eval_metrics,
+              eval_py_env,
+              eval_py_policy,
+              num_episodes=num_eval_episodes,
+              global_step=global_step_val,
+              callback=eval_metrics_callback,
+          )
+
+
+def main(_):
```

Comment:
Branching from the V2 and using eager mode in general will probably make your life debugging a lot easier.

Reply:
I can already see this is making a big difference! Thank you for the suggestion! :+1: